p8e-api/src/main/kotlin/io/provenance/engine/service/ProvenanceGrpcService.kt

```diff
@@ -0,0 +1,147 @@
+package io.provenance.engine.service
+
+import com.google.protobuf.Any
+import com.google.protobuf.Message
+import cosmos.auth.v1beta1.Auth
+import cosmos.auth.v1beta1.QueryGrpc
+import cosmos.auth.v1beta1.QueryOuterClass
+import cosmos.base.v1beta1.CoinOuterClass
+import cosmos.crypto.secp256k1.Keys
+import cosmos.tx.signing.v1beta1.Signing
+import cosmos.tx.v1beta1.ServiceGrpc
+import cosmos.tx.v1beta1.ServiceOuterClass.*
+import cosmos.tx.v1beta1.TxOuterClass.*
+import io.grpc.netty.NettyChannelBuilder
+import io.p8e.util.toByteString
+import io.provenance.engine.config.ChaincodeProperties
+import io.provenance.engine.crypto.Account
+import io.provenance.engine.crypto.PbSigner
+import io.provenance.metadata.v1.p8e.ContractSpec
+import io.provenance.pbc.clients.roundUp
+import org.kethereum.crypto.getCompressedPublicKey
+import org.springframework.stereotype.Service
+import java.net.URI
+import java.util.concurrent.TimeUnit
+import java.util.logging.Level
+import java.util.logging.Logger
+import io.p8e.proto.ContractSpecs.ContractSpec as P8EContractSpec
+
+@Service
+class ProvenanceGrpcService(
+    private val accountProvider: Account,
+    private val chaincodeProperties: ChaincodeProperties
+) {
+    private val channel =  URI(chaincodeProperties.url).let { uri ->
+            Logger.getLogger("io.netty").setLevel(Level.ALL)
+            NettyChannelBuilder.forAddress(uri.host, uri.port)
+                .also {
+                    if (uri.scheme == "grpcs") {
+                        it.useTransportSecurity()
+                    } else {
+                        it.usePlaintext()
+                    }
+                }
+                .maxInboundMessageSize(40 * 1024 * 1024) // ~ 20 MB
+                .idleTimeout(5, TimeUnit.MINUTES)
+                .keepAliveTime(60, TimeUnit.SECONDS) // ~ 12 pbc block cuts
+                .keepAliveTimeout(20, TimeUnit.SECONDS)
+                .build()
+        }
+
+    private val cosmosService = ServiceGrpc.newBlockingStub(channel)
+    private val accountService = QueryGrpc.newBlockingStub(channel)
+
+    private val bech32Address = accountProvider.bech32Address()
+    private val keyPair = accountProvider.getKeyPair()
+    private val signer = PbSigner.signerFor(keyPair)
+
+    fun accountInfo(): Auth.BaseAccount = accountService.account(QueryOuterClass.QueryAccountRequest.newBuilder()
+            .setAddress(bech32Address)
+            .build()
+        ).run { account.unpack(Auth.BaseAccount::class.java) }
+
+    fun signTx(body: TxBody, accountInfo: Auth.BaseAccount, sequenceNumberOffset: Int = 0, gasEstimate: GasEstimate = GasEstimate(0)): Tx {
```

Comment:
Note: I haven't tested the sequence number and offset, once that is all converted though for the memorialize contract, I would expect it to work fine if you stuff a bunch of transactions inside a block following the logic we currently use with SimpleClient

Reply:
In case this comes up... one signed tx (sequence number) with many messages... the terminology can get confusing on the blockchain side because we think about submitting many requests as many transactions... this point becomes important though because all those messages succeed or fail as a group. You can submit more than one tx per block (potentially again with more than one msg each) if you bump the sequence number each time you sign.